server/src/main/java/org/elasticsearch/search/runtime/IpScriptFieldTermsQuery.java

```diff
@@ -26,9 +26,11 @@ public IpScriptFieldTermsQuery(Script script, IpFieldScript.LeafFactory leafFact
 
     @Override
     protected boolean matches(BytesRef[] values, int count) {
-        for (int i = 0; i < count; i++) {
-            if (terms.find(values[i]) >= 0) {
-                return true;
+        synchronized (terms) {
```

Comment:
OK, I recognize this issue now. It seems like something similar here: https://github.com/elastic/elasticsearch/pull/106678/files

I think this is because we are searching multiple segments in parallel. But why are we seeing weird failures here but not in `IpScriptFieldTermQuery`? 

@javanna you might be interested here.

Reply:
> I think this is because we are searching multiple segments in parallel

It definitely looks like it. Would this PR be the proper way of fixing this issue, or should we look into something different?

> But why are we seeing weird failures here but not in IpScriptFieldTermQuery?

`IpScriptFieldTermQuery` does use a `BytesRef` instead of a `BytesRefHash`, checking the value directly instead of using the hash. So this is thread safe as opposed to using the `BytesRefHash` from multiple threads.

Reply:
I don't think the issue is the `BytesRefHash` looking at how its used, we construct one new `BytesRefHash`, statically setting all the terms that will match and then pass it in. We don't reuse buffers or anything. 

Do you see something where we are adjusting that bytes ref hash? Or is the internal hash mutated due to find checks?

Reply:
Yes, this puzzled me for a while as well. I think the problem is located [here](https://github.com/elastic/elasticsearch/blob/aee3a253dd8f5c40be9d3b1686b8c7889556a2f3/server/src/main/java/org/elasticsearch/common/util/BytesRefHash.java#L139) - `spare` is a class attribute that is being (re)used for performing the `find` operation.

This aligns with the `find` operation being nondeterministic in my tests, and thus being unable to find values that were in the hash.

Reply:
Oh wow, `spare` is indeed a huge issue. What on Earth!!!

Reply:
I wonder if we should just use Lucenes BytesRefHash here instead of using ours which from what I can tell is only useful if we are actually using BigArrays for reuse, and here, we are not.

Reply:
I think the lock here is indeed frustrating. Another option is extending `BytesRefHash` that allows to return a "readonly" threadsafe version where `find` is only satisfied so that we can use a `BytesRefHash.getFinder` and each `finder` has its own `spare` and it relies on the original bytes ref hash.

Reply:
I like the idea of making a threadsafe version of `BytesRefHash`. I guess the main goal of `spare` is avoiding repeated allocations for it.

> Another option is extending BytesRefHash that allows to return a "readonly" threadsafe version where find is only satisfied so that we can use a BytesRefHash.getFinder and each finder has its own spare and it relies on the original bytes ref hash.

How would we create a separate finder for each thread that is accessing the query `matches` method?

What do you think about using a `ThreadLocal `variable for holding `spare` instances in a threadsafe version of `BytesRefHash`?

Reply:
@carlosdelest I would override the `createTwoPhaseIterator` method. For `BytesRefHash` I would create a method that returns a inner class `BytesRefHash.Finder` or something that uses its own `spare` when searching the parent `BytesRefHash` class and then use that `Finder` inside the `createTwoPhaseIterator` that calls `matches`.